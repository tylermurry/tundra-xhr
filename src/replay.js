/* eslint-disable no-underscore-dangle */

import 'url';
import xhrMock from 'xhr-mock';
import omit from 'lodash.omit';
import buildRequestRepeatMap from './requestRepeatMapBuilder';
import stringIsSimilarTo from './stringSimilarity';
import buildRequestId from './requestIdBuilder';
import removeURLPrefix from './removeURLPrefix';
import submitRequestData from './submitRequest';
import buildRepeatableMockFunction from './repeatableMockFunctionBuilder';
import buildRequest from './requestBuilder';

const DEFAULT_CONFIG = {
  debuggingEnabled: true,
  debugPort: 9091,
};

export const requestMatches = (matchingConfig, profileRequest, request) => {
  const headersToOmit = matchingConfig ? matchingConfig.headersToOmit : null;
  const requestHeaders = JSON.stringify(omit(request._headers, headersToOmit));
  const profileRequestHeaders = JSON.stringify(omit(profileRequest.headers, headersToOmit));

  const urlMatches = stringIsSimilarTo(removeURLPrefix(profileRequest.url), removeURLPrefix(request._url.toString()));
  const bodyMatches = request._method !== 'GET' ? stringIsSimilarTo(profileRequest.content, request._body) : true;
  const headersMatch = stringIsSimilarTo(profileRequestHeaders, requestHeaders);
  const methodMatches = request._method === profileRequest.method;

  return urlMatches && methodMatches && bodyMatches && headersMatch;
};

export const matchingFunction = (matchingConfig, requestRepeatMap, request, response, realXHR) => {
  const repeatableMockFunction = buildRepeatableMockFunction(matchingConfig, requestRepeatMap, request, response);

  return (req, res) => {
    const everythingMatches = requestMatches(matchingConfig, request, req);

    if (matchingConfig && matchingConfig.debuggingEnabled) {
      submitRequestData(
        buildRequest(req, res),
        matchingConfig.debugPort,
        everythingMatches,
        realXHR,
      );
    }

    if (everythingMatches) {
      return repeatableMockFunction(req, res);
    }
    return false;
  };
};

export default (profileRequests, config) => {
  const realXHR = global.XMLHttpRequest;

  xhrMock.setup();
  xhrMock.reset();

  const defaultedConfig = { ...DEFAULT_CONFIG, ...config };
  const repeatMap = buildRequestRepeatMap(profileRequests);

  profileRequests.forEach(({ request, response }) => {
    const requestRepeatMap = repeatMap[buildRequestId(request)];
    requestRepeatMap.invocations += 1;

    xhrMock.use(
      request.method.toLowerCase(),
      new RegExp('.*'),
      matchingFunction(defaultedConfig, repeatMap, request, response, realXHR),
    );
  });
};
