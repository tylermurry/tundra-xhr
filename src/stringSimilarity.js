import escapeRegExp from 'lodash.escaperegexp';
import matcher from 'matcher';

export const WILDCARD_MARKER_ESCAPED = '{{\\*}}';
export const WILDCARD_MARKER = '{{*}}';

export default (source, target) => {
  if (!source || (source || '') === (target || '')) {
    return source === target;
  }

  const wildcardedSource = source
    .replace(new RegExp(escapeRegExp('*'), 'g'), '\\*')
    .replace(new RegExp(escapeRegExp(WILDCARD_MARKER_ESCAPED), 'g'), '*');

  const escapedTarget = target
    ? target.replace(new RegExp(escapeRegExp('*'), 'g'), '\\*')
    : null;

  return matcher.isMatch(escapedTarget, wildcardedSource);
};
