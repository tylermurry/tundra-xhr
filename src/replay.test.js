import xhrMock from 'xhr-mock';
import replay, { matchingFunction, requestMatches } from './replay';
import { WILDCARD_MARKER } from './stringSimilarity';
import submitRequestData from './submitRequest';
import buildRepeatableMockFunction from './repeatableMockFunctionBuilder';

const emptyProfile = require('./fixtures/profiles/no-requests');
const singleRequest = require('./fixtures/profiles/single-request');
const multipleRequests = require('./fixtures/profiles/multiple-requests');

jest.mock('xhr-mock', () => ({
  reset: jest.fn(),
  setup: jest.fn(),
  use: jest.fn(),
}));

jest.mock('./submitRequest', () => jest.fn());
jest.mock('./repeatableMockFunctionBuilder', () => jest.fn());

describe('replay', () => {
  let profileRequest;
  let request;
  let response;

  const baseProfileRequest = {
    method: 'GET',
    url: 'http://www.someurl.com',
    headers: {
      abc: ['123'],
      xyz: ['456'],
    },
    content: 'body',
  };

  const baseRequest = {
    _url: 'http://www.someurl.com',
    _method: 'GET',
    _headers: {
      abc: ['123'],
      xyz: ['456'],
    },
    _body: 'body',
  };

  const baseResponse = {
    _status: 200,
    _headers: {
      abc: ['321'],
    },
    _body: 'body1',
  };

  describe('requestMatches', () => {
    beforeEach(() => {
      jest.resetAllMocks();

      profileRequest = JSON.parse(JSON.stringify(baseProfileRequest));
      request = JSON.parse(JSON.stringify(baseRequest));
      response = JSON.parse(JSON.stringify(baseResponse));
    });

    it('should match a standard request on all factors and an empty matching config', () => {
      expect(requestMatches({}, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors and a null matching config', () => {
      expect(requestMatches(null, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors and a matching config with headers to omit', () => {
      profileRequest.headers.xyz = 'something else';
      expect(requestMatches({ headersToOmit: ['xyz'] }, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors and a matching config with debugging enabled', () => {
      const config = { debuggingEnabled: true, debugPort: 9091 };

      expect(requestMatches(config, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors with a fuzzy url', () => {
      profileRequest.url = `http://www.${WILDCARD_MARKER}.com`;
      expect(requestMatches({}, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors with a fuzzy header', () => {
      profileRequest.headers.abc = [`1${WILDCARD_MARKER}3`];
      expect(requestMatches({}, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should match a standard request on all factors with a fuzzy body', () => {
      profileRequest.body = `b${WILDCARD_MARKER}y`;
      expect(requestMatches({}, profileRequest, request)).toBe(true);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should not match a standard request because the URL doesn\'t match', () => {
      profileRequest.url = 'bad';
      expect(requestMatches({}, profileRequest, request)).toBe(false);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should not match a standard request because the headers don\'t match', () => {
      profileRequest.headers.abc = 'bad';
      expect(requestMatches({}, profileRequest, request)).toBe(false);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should not match a standard request because the method doesn\'t match', () => {
      profileRequest.method = 'bad';
      expect(requestMatches({}, profileRequest, request)).toBe(false);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should not match a standard request because the body doesn\'t match', () => {
      profileRequest.method = 'POST';
      profileRequest.content = 'bad';
      expect(requestMatches({}, profileRequest, request)).toBe(false);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });
  });

  describe('matchingFunction', () => {
    beforeEach(() => {
      jest.resetAllMocks();

      profileRequest = JSON.parse(JSON.stringify(baseProfileRequest));
      request = JSON.parse(JSON.stringify(baseRequest));
      response = JSON.parse(JSON.stringify(baseResponse));
    });

    it('should return the correct mock function when everything matches and debugging is enabled', () => {
      buildRepeatableMockFunction.mockImplementation(() => () => 'correct mock function');

      const mockFunction = matchingFunction({ debuggingEnabled: true }, null, profileRequest, null, null);

      expect(mockFunction(request, response)).toBe('correct mock function');
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should provide the correct mock function when everything matches and debugging is disabled', () => {
      buildRepeatableMockFunction.mockImplementation(() => () => 'correct mock function');

      const mockFunction = matchingFunction({ }, null, profileRequest, null, null);

      expect(mockFunction(request, response)).toBe('correct mock function');
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });

    it('should provide the mocking function that returns false when something doesn\'t match', () => {
      const mockFunction = matchingFunction({ }, null, profileRequest, null, null);

      profileRequest.url = 'something different';

      expect(mockFunction(request, response)).toBe(false);
      expect(submitRequestData.mock.calls).toMatchSnapshot();
    });
  });

  describe('default', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    it('should mock requests for an empty profile', () => {
      replay(emptyProfile, {});

      expect(xhrMock.reset).toBeCalled();
      expect(xhrMock.setup).toBeCalled();
      expect(xhrMock.use.mock.calls).toEqual([]);
    });

    it('should mock requests for a profile with a single request', () => {
      replay(singleRequest, {});

      expect(xhrMock.reset).toBeCalled();
      expect(xhrMock.setup).toBeCalled();
      expect(xhrMock.use.mock.calls).toMatchSnapshot();
    });

    it('should mock requests for a profile with two requests and the default repeat mode', () => {
      replay(multipleRequests, {});

      expect(xhrMock.reset).toBeCalled();
      expect(xhrMock.setup).toBeCalled();
      expect(xhrMock.use.mock.calls).toMatchSnapshot();
    });

    it('should mock requests for a profile with two requests and a repeat mode of \'first\'', () => {
      replay(multipleRequests, { repeatMode: 'first' });

      expect(xhrMock.reset).toBeCalled();
      expect(xhrMock.setup).toBeCalled();
      expect(xhrMock.use.mock.calls).toMatchSnapshot();
    });

    it('should mock requests for a profile with two requests and a repeat mode of \'last\'', () => {
      replay(multipleRequests, { repeatMode: 'last' });

      expect(xhrMock.reset).toBeCalled();
      expect(xhrMock.setup).toBeCalled();
      expect(xhrMock.use.mock.calls).toMatchSnapshot();
    });
  });
});
