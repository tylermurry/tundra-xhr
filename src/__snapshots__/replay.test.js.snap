// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replay default should mock requests for a profile with a single request 1`] = `
Array [
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
]
`;

exports[`replay default should mock requests for a profile with two requests and a repeat mode of 'first' 1`] = `
Array [
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
]
`;

exports[`replay default should mock requests for a profile with two requests and a repeat mode of 'last' 1`] = `
Array [
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
]
`;

exports[`replay default should mock requests for a profile with two requests and the default repeat mode 1`] = `
Array [
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
  Array [
    "post",
    /\\.\\*/,
    [Function],
  ],
]
`;

exports[`replay matchingFunction should provide the correct mock function when everything matches and debugging is disabled 1`] = `Array []`;

exports[`replay matchingFunction should provide the mocking function that returns false when something doesn't match 1`] = `Array []`;

exports[`replay matchingFunction should return the correct mock function when everything matches and debugging is enabled 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "content": "body",
        "headers": Object {
          "abc": Array [
            "123",
          ],
          "xyz": Array [
            "456",
          ],
        },
        "method": "GET",
        "url": "http://www.someurl.com",
      },
      "response": Object {
        "content": "body1",
        "headers": Object {
          "abc": Array [
            "321",
          ],
        },
        "statusCode": 200,
      },
    },
    undefined,
    true,
    null,
  ],
]
`;

exports[`replay requestMatches should match a standard request on all factors and a matching config with debugging enabled 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors and a matching config with headers to omit 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors and a null matching config 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors and an empty matching config 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors with a fuzzy body 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors with a fuzzy header 1`] = `Array []`;

exports[`replay requestMatches should match a standard request on all factors with a fuzzy url 1`] = `Array []`;

exports[`replay requestMatches should not match a standard request because the URL doesn't match 1`] = `Array []`;

exports[`replay requestMatches should not match a standard request because the body doesn't match 1`] = `Array []`;

exports[`replay requestMatches should not match a standard request because the headers don't match 1`] = `Array []`;

exports[`replay requestMatches should not match a standard request because the method doesn't match 1`] = `Array []`;
